### DISCUSSION TITLE
How to wrap components that use the getLayout() method

### DISCUSSION THREAD

**Discussion by noahlst** [vercel/next.js#36725] [Help] [answered]
How can I wrap components that use the getLayout() method?

Take the example below:

```js
export default function Application({ Component, pageProps }) {
  const getLayout = Component.getLayout ?? ((page) => page);

  // Render the children with layout persistence
  return getLayout(<Component {...pageProps} />);
}
```

This is taken from [the "Layouts" Next.js docs](https://nextjs.org/docs/basic-features/layouts).

How can I wrap the `<Component />`? I need to implement provider components that apply to the whole app (`<ChakraProvider />`).

This is what I implemented:

```js
export default function Application({ Component, pageProps }) {
  const getLayout = Component.getLayout ?? ((page) => page);

  // Render the children with layout persistence
  function ApplicationCore() {
    return getLayout(<Component {...pageProps} />);
  }

  return (
    <ChakraProvider theme={theme}>
      <ApplicationCore />
    </ChakraProvider>
  );
}
```

This works, but the layout persistence simply doesn't work. It removes all benefits from the first example.

Is it possible to fix this?

**Comment by icyJoseph** [comment, marked as answer]
Could you try with: 

```js
export default function Application({ Component, pageProps }) {
  const getLayout = Component.getLayout ?? ((page) => page);

  return (
    <ChakraProvider theme={theme}>
      {getLayout(<Component {...pageProps} />)}
    </ChakraProvider>
  );
}
```

**Reply by icyJoseph** [reply]
The long winded explanation is that you are destroying the component on every render. Please see here, https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#component-types-and-reconciliation. Most notably:

> This means that you must never create new component types while rendering! Whenever you create a new component type, it's a different reference, and that will cause React to repeatedly destroy and recreate the child component tree.

```jsx
function ParentComponent() {
  // This creates a new `ChildComponent` reference every time!
  function ChildComponent() {}
  
  return <ChildComponent />
}
```

**Reply by noahlst** [reply]
@icyJoseph This works. Thanks for the help!
